utils_test.go

```diff
@@ -293,44 +294,184 @@ func TestIsolationLevelMapping(t *testing.T) {
+			str:  "2020-05-25 23:22:01.15949",
+		},
+		{
+			name: "parse datetime nanosec 6-digits",
+			str:  "2020-05-25 23:22:01.159491",
+		},
+	}
+
+	for _, loc := range []*time.Location{
+		time.UTC,
+		time.FixedZone("test", 8*60*60),
+	} {
+		for _, cc := range cases {
+			t.Run(cc.name+"-"+loc.String(), func(t *testing.T) {
+				want, err := deprecatedParseDateTime(cc.str, loc)
+				if err != nil {
+					t.Fatal(err)
+				}
+				got, err := parseDateTime([]byte(cc.str), loc)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				if !want.Equal(got) {
+					t.Fatalf("want: %v, but got %v", want, got)
+				}
+			})
 		}
 	}
 }
 
+func TestParseDateTimeFail(t *testing.T) {
+	cases := []struct {
+		name    string
+		str     string
+		wantErr string
+	}{
+		{
+			name:    "parse invalid time",
+			str:     "hello",
+			wantErr: "invalid time bytes: hello",
+		},
+		{
+			name:    "parse year",
+			str:     "000!-00-00 00:00:00.000000",
+			wantErr: "not [0-9]",
+		},
+		{
+			name:    "parse month",
+			str:     "0000-!0-00 00:00:00.000000",
+			wantErr: "not [0-9]",
+		},
+		{
+			name:    `parse "-" after parsed year`,
+			str:     "0000:00-00 00:00:00.000000",
+			wantErr: "bad value for field: `:`",
+		},
+		{
+			name:    `parse "-" after parsed month`,
+			str:     "0000-00:00 00:00:00.000000",
+			wantErr: "bad value for field: `:`",
+		},
+		{
+			name:    `parse " " after parsed date`,
+			str:     "0000-00-00+00:00:00.000000",
+			wantErr: "bad value for field: `+`",
+		},
+		{
+			name:    `parse ":" after parsed date`,
+			str:     "0000-00-00 00-00:00.000000",
+			wantErr: "bad value for field: `-`",
+		},
+		{
+			name:    `parse ":" after parsed hour`,
+			str:     "0000-00-00 00:00-00.000000",
+			wantErr: "bad value for field: `-`",
+		},
+		{
+			name:    `parse "." after parsed sec`,
+			str:     "0000-00-00 00:00:00?000000",
+			wantErr: "bad value for field: `?`",
+		},
+	}
+
+	for _, cc := range cases {
+		t.Run(cc.name, func(t *testing.T) {
+			got, err := parseDateTime([]byte(cc.str), time.UTC)
+			if err == nil {
+				t.Fatal("want error")
+			}
+			if cc.wantErr != err.Error() {
+				t.Fatalf("want `%s`, but got `%s`", cc.wantErr, err)
+			}
+			if !got.IsZero() {
+				t.Fatal("want zero time")
+			}
+		})
+	}
+}
+
 func BenchmarkParseDateTime(b *testing.B) {
 	str := "2020-05-13 21:30:45"
 	loc := time.FixedZone("test", 8*60*60)
 	for i := 0; i < b.N; i++ {
-		_, _ = parseDateTime(str, loc)
+		_, _ = deprecatedParseDateTime(str, loc)
+	}
+}
+
+func BenchmarkParseByteDateTime(b *testing.B) {
+	bStr := []byte("2020-05-25 23:22:01.159491")
+	loc := time.FixedZone("test", 8*60*60)
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		_, _ = parseDateTime(bStr, loc)
+	}
+}
+
+func BenchmarkParseByteDateTimeStringCast(b *testing.B) {
+	bStr := []byte("2020-05-25 23:22:01.159491")
+	loc := time.FixedZone("test", 8*60*60)
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		_, _ = deprecatedParseDateTime(string(bStr), loc)
```

Comment:
Please post the result of this benchmark and remove `deprecatedParseDateTime`.
We do not keep old implementations only for benchmark.

Reply:
removed https://github.com/go-sql-driver/mysql/pull/1113/commits/2c6aa27e88345456be824ebb72dd476cf69506c4